src/nfa/nfa.cc

```diff
@@ -1052,28 +1131,61 @@ Nfa::Nfa::StateBoolArray Nfa::compute_reachability() const
     return reachable;
 }
 
-Nfa Nfa::get_digraph()
-{
-    Nfa digraph{ get_num_of_states(), initialstates, finalstates};
-    Symbol abstract_symbol{ 'x' };
 
-    for (State src_state{ 0 }; src_state < get_num_of_states(); ++src_state)
-    {
-        for (const auto &symbol_transitions: this->transitionrelation[src_state])
-        {
-            for (State tgt_state: symbol_transitions.states_to)
-            {
-                if (!digraph.has_trans(src_state, abstract_symbol, tgt_state))
-                {
-                    digraph.add_trans(src_state, abstract_symbol, tgt_state);
+Nfa::StateBoolArray Nfa::compute_reachability(const Nfa::StateBoolArray& states_to_consider) const {
+    std::vector<State> worklist{};
+    StateBoolArray reachable(get_num_of_states(), false);
+    for (const State state: initialstates) {
+        if (states_to_consider[state]) {
+            worklist.push_back(state);
+            reachable.at(state) = true;
+        }
+    }
+
+    State state{};
+    while (!worklist.empty()) {
+        state = worklist.back();
+        worklist.pop_back();
+
+        for (const auto& state_transitions: transitionrelation[state]) {
+            for (const State target_state: state_transitions.states_to) {
+                if (states_to_consider[target_state] && !reachable[target_state]) {
+                    worklist.push_back(target_state);
+                    reachable[target_state] = true;
                 }
             }
         }
     }
 
+    return reachable;
+}
+
+Nfa Nfa::get_digraph() const {
+    Nfa digraph{ get_num_of_states(), initialstates, finalstates};
+    collect_directed_transitions(digraph);
     return digraph;
 }
 
+void Nfa::get_digraph(Nfa& result) const {
+    result = get_digraph();
+}
+
+void Nfa::collect_directed_transitions(Nfa& digraph) const {
+    constexpr Symbol abstract_symbol{ 'x' };
```

Comment:
This could be somewhere "higher" in the hierarchy. What if the abstract symbol changes? One would have to find it in this function? Wouldn't it make sense to have it e.g. as class constant usable in other places?

Reply:
This is a temporary solution until we figure out how to handle alphabets, epsilon(s), complement etc. I thought we could use some “special” symbol that would be in general used for similar situations. If we have reserved values for special symbols and/or epsilon symbols, for example, we could use these so that the abstract symbol would not be some random generally usable symbol.

Alternatively, we could handle transitions with `do not care` symbols and/or epsilon symbols separately, maybe. But when this was discussed earlier, we have agreed that we do not want to modify the current way we handle transitions.

Reply:
For now, we could have an optional parameter for `get_digraph` where the user can specify the abstract symbol themself. By default, the symbol would be, for example, this temporary `'x'`.

Reply:
I have implemented this. Do we want to introduce a `Nfa`-wide constant possibly usable elsewhere (where?) instead?